runtime_scan/pkg/orchestrator/scanwatcher/watcher.go

```diff
@@ -0,0 +1,220 @@
+	client           *backendclient.BackendClient
+	pollPeriod       time.Duration
+	reconcileTimeout time.Duration
+}
+
+func (w *Watcher) Start(ctx context.Context) {
+	queue := common.NewQueue[ScanReconcileEvent]()
+
+	poller := &ScanPoller{
+		Logger:     w.logger,
+		PollPeriod: w.pollPeriod,
+		Queue:      queue,
+		GetItems:   w.GetAbortedScans,
+	}
+	poller.Start(ctx)
+
+	reconciler := &ScanReconciler{
+		ReconcileTimeout:  w.reconcileTimeout,
+		Queue:             queue,
+		ReconcileFunction: w.Reconcile,
+	}
+	reconciler.Start(ctx)
+}
+
+func (w *Watcher) GetAbortedScans(ctx context.Context) ([]ScanReconcileEvent, error) {
+	scans, err := w.getScansByState(ctx, models.ScanStateAborted)
+	if err != nil || scans.Items == nil || len(*scans.Items) <= 0 {
+		return nil, err
+	}
+
+	count := len(*scans.Items)
+	r := make([]ScanReconcileEvent, count)
+	for i, scan := range *scans.Items {
+		r[i] = ScanReconcileEvent{
+			ScanID: *scan.Id,
+		}
+	}
+
+	return r, nil
+}
+
+func (w *Watcher) Reconcile(ctx context.Context, event ScanReconcileEvent) error {
+	w.logger.Infof("reconciling scan event: %v", event)
+
+	selector := "id,state,stateReason"
+	params := models.GetScansScanIDParams{
+		Select: &selector,
+	}
+
+	scan, err := w.client.GetScan(ctx, event.ScanID, params)
+	if err != nil || scan == nil {
+		return fmt.Errorf("getting scan with id %s failed: %v", event.ScanID, err)
+	}
+
+	state, ok := scan.GetState()
+	if !ok {
+		return fmt.Errorf("cannot determine state of Scan with %s id", event.ScanID)
+	}
+
+	switch state {
+	case models.ScanStateDone, models.ScanStateFailed:
+		w.logger.Infof("reconciling scan event is skipped as Scan is already finished: %v", event)
+	case models.ScanStateAborted:
+		return w.reconcileAborted(ctx, event)
+	case models.ScanStatePending, models.ScanStateDiscovered, models.ScanStateInProgress:
+		fallthrough
+	default:
+	}
+
+	return nil
+}
+
+func (w *Watcher) getScansByState(ctx context.Context, s models.ScanState) (models.Scans, error) {
+	filter := fmt.Sprintf("state eq '%s'", s)
+	selector := "id"
+	params := models.GetScansParams{
+		Filter: &filter,
+		Select: &selector,
+	}
+	scans, err := w.client.GetScans(ctx, params)
+	if err != nil {
+		err = fmt.Errorf("getting Scan(s) by their state failed: %v", err)
+	}
+	if scans == nil {
+		scans = &models.Scans{}
+	}
+
+	return *scans, err
+}
+
+func (w *Watcher) reconcileAborted(ctx context.Context, event ScanReconcileEvent) error {
+	filter := fmt.Sprintf("scan/id eq '%s' and status/general/state ne '%s' and status/general/state ne '%s'",
+		event.ScanID, models.ABORTED, models.DONE)
+	selector := "id,status"
+	params := models.GetScanResultsParams{
+		Filter: &filter,
+		Select: &selector,
+	}
+
+	scanResults, err := w.client.GetScanResults(ctx, params)
+	if err != nil {
+		return fmt.Errorf("getting ScanResult(s) for Scan with %s id failed: %v", event.ScanID, err)
+	}
+
+	if scanResults.Items != nil && len(*scanResults.Items) > 0 {
```

Comment:
maybe better to return if scanResults.Items == nil || len(*scanResults.Items) == 0

Reply:
Nope. We still need to update the Scan [here](https://github.com/openclarity/vmclarity/pull/218/files#diff-ffbf779aa4010b4a63a4c4576e2765b61ba56f7db4d4249002b9e27f037ab4e6R206-R217) which is temporarily disabled as it conflicts with the code in `Scanner.jobBatchManagement`. I've started to refactor the whole Scan lifecycle management by adopting the reconciler logic and moving Scan related operations to `ScanWatcher`.

So I'd keep it as is.